### PR TITLE
Document unsafety in slice/sort.rs

### PR DESCRIPTION
**Pull request by hbina**
Let me know if these documentations are accurate c:

I don't think I am capable enough to document the safety of `partition_blocks`, however.

Related issue #66219

### CODE REVIEW

File: src/libcore/slice/sort.rs [hunk 1]
```diff
@@ -32,6 +32,20 @@ where
...
+    //
+    // a. Indexing:
+    //  1. We checked the size of the array to >=2.
+    //  2. All the indexing that we will do is always between {0 <= index < len} at most.
+    //
+    // b. Memory copying
+    //  1. We are obtaining pointers to references which are guaranteed to be valid.
+    //  2. They cannot overlap because we obtain pointers to difference indices of the slice.
+    //     Namely, `i` and `i-1`.
+    //  3. FIXME: Guarantees that the elements are properly aligned?
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:46]: If the slice is properly aligned, the elements will be properly aligned. And it's the caller's responsibility to make sure the slice is properly aligned.

File: src/libcore/slice/sort.rs [hunk 2]
```diff
@@ -62,6 +76,20 @@ where
...
+    //
+    // a. Indexing:
+    //  1. We checked the size of the array to >=2.
+    //  2. All the indexing that we will do is always between {0 <= index < len-1} at most.
+    //
+    // b. Memory copying
+    //  1. We are obtaining pointers to references which are guaranteed to be valid.
+    //  2. They cannot overlap because we obtain pointers to difference indices of the slice.
+    //     Namely, `i` and `i+1`.
+    //  3. FIXME: Guarantees that the elements are properly aligned?
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:90]: Same comment as above.

File: src/libcore/slice/sort.rs [hunk 3]
```diff
@@ -103,6 +131,8 @@ where
     let mut i = 1;
 
     for _ in 0..MAX_STEPS {
+        // SAFETY: We already explicitly done the bound checking with `i<len`
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:134]:
```suggestion
        // SAFETY: We already explicitly did the bound checking with `i<len`
```

File: src/libcore/slice/sort.rs [hunk 4]
```diff
@@ -103,6 +131,8 @@ where
...
+        // All our indexing following that is only in the range {0 <= index < len}
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:135]:
```suggestion
        // All our subsequent indexing is only in the range `0 <= index < len`
```

File: src/libcore/slice/sort.rs [hunk 5]
```diff
@@ -62,6 +76,20 @@ where
     F: FnMut(&T, &T) -> bool,
 {
     let len = v.len();
+    // SAFETY: As with shift_head, the unsafe operations below involves indexing without a bound check (`get_unchecked` and `get_unchecked_mut`)
+    // and copying memory (`ptr::copy_nonoverlapping`).
+    //
+    // a. Indexing:
+    //  1. We checked the size of the array to >=2.
+    //  2. All the indexing that we will do is always between {0 <= index < len-1} at most.
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:84]:
```suggestion
    //  2. All the indexing that we will do is always between `0 <= index < len-1` at most.
```

File: src/libcore/slice/sort.rs [hunk 6]
```diff
@@ -404,8 +435,12 @@ where
         // Find the first pair of out-of-order elements.
         let mut l = 0;
         let mut r = v.len();
+
+        // SAFETY: The unsafety below involves indexing an array.
+        // For the first one: we already do the bound checking here with `l<r`.
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:440]:
```suggestion
        // For the first one: we already do the bounds checking here with `l < r`.
```

File: src/libcore/slice/sort.rs [hunk 7]
```diff
@@ -404,8 +435,12 @@ where
...
+        // For the secondn one: the minimum value for `l` is 0 and the maximum value for `r` is `v.len().`
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:441]:
```suggestion
        // For the second one: the minimum value for `l` is 0 and the maximum value for `r` is `v.len().`
```

File: src/libcore/slice/sort.rs [hunk 8]
```diff
@@ -452,8 +488,11 @@ where
     let mut l = 0;
     let mut r = v.len();
     loop {
+        // SAFETY: The unsafety below involves indexing an array.
+        // For the first one: we already do the bound checking here with `l<r`.
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:492]:
```suggestion
        // For the first one: we already do the bounds checking here with `l < r`.
```

**Review by joshtriplett — CHANGES_REQUESTED** [review]
I posted a few review comments; with those fixed, r=me.

**Comment by hbina** [comment]
Thanks @joshtriplett ! I have applied your requested changes and added some more safety comments. I read through the discussions by more senior people over at https://github.com/rust-lang/rust/pull/63793
I think some of my documentations might have been too verbose.
Lemme know if they are correct and/or concise enough.

**Comment by joshtriplett** [comment]
Looks good to me (and definitely not too verbose). However, it looks like the build is failing because you removed `ignore-tidy-undocumented-unsafe` but there are still some undocumented unsafe blocks. You don't have to document them, but to keep the build passing, can you leave that comment in please?

File: src/libcore/slice/sort.rs [hunk 9]
```diff
@@ -6,12 +6,12 @@
 //! Unstable sorting is compatible with libcore because it doesn't allocate memory, unlike our
 //! stable sorting implementation.
 
-// ignore-tidy-undocumented-unsafe
-
 use crate::cmp;
 use crate::mem::{self, MaybeUninit};
 use crate::ptr;
 
+// ignore-tidy-undocumented-unsafe
```
Comment (joshtriplett) [inline src/libcore/slice/sort.rs:13]: Please don't move this comment below the `use` lines like this. See the overall diff; these two lines shouldn't change at all.

**Pull request merged**